sklearn/preprocessing/_function_transformer.py

```diff
@@ -156,7 +156,16 @@ def _check_inverse_transform(self, X):
         """Check that func and inverse_func are the inverse."""
         idx_selected = slice(None, None, max(1, X.shape[0] // 100))
         X_round_trip = self.inverse_transform(self.transform(X[idx_selected]))
-        if not _allclose_dense_sparse(X[idx_selected], X_round_trip):
+
+        X = np.asarray(X) if not hasattr(X, "dtype") else X
+        if np.issubdtype(X.dtype, np.number):
+            valid = _allclose_dense_sparse(X[idx_selected], X_round_trip)
+        else:
+            raise ValueError(
+                "'check_inverse' is only supported when all the elements in `X` is"
+                " numerical."
+            )
+        if not valid:
```

Comment:
I think we can early exit here:

```python
        if not np.issubdtype(X.dtype, np.number):
            raise ValueError(
                "'check_inverse' is only supported when all the elements in `X` is"
                " numerical."
            )

        if not _allclose_dense_sparse(X[idx_selected], X_round_trip):
            warnings.warn(
                "The provided functions are not strictly"
                " inverse of each other. If you are sure you"
                " want to proceed regardless, set"
                " 'check_inverse=False'.",
                UserWarning,
            )
```

Also we do not need to cast since this function already assumes that `X` is an array or sparse. (There are calls to `X.shape[0]` and at the beginning of `_check_inverse_transform`.